vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/dao/impl/DeviceRepositoryImpl.java

```diff
@@ -0,0 +1,175 @@
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package org.opensds.vmware.ngc.dao.impl;
+
+import org.opensds.vmware.ngc.common.Storage;
+import org.opensds.vmware.ngc.common.StorageFactory;
+import org.opensds.vmware.ngc.models.StorageMO;
+import org.opensds.vmware.ngc.dao.DeviceRepository;
+import org.opensds.vmware.ngc.model.DeviceInfo;
+import org.opensds.vmware.ngc.util.Constant;
+import org.opensds.vmware.ngc.util.FileManager;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Repository;
+
+import javax.annotation.PostConstruct;
+import java.io.*;
+import java.util.Map;
+import java.util.concurrent.*;
+
+@Repository
+@Lazy(value = false)
+public class DeviceRepositoryImpl implements DeviceRepository {
+
+    private static final Log logger = LogFactory.getLog(DeviceRepositoryImpl.class);
+    private static Map<String, DeviceInfo> CACHE_DEVICE_INFO = new ConcurrentHashMap<>();
+    private static Map<String, Storage> LOGINED_DEVICE = new ConcurrentHashMap<>();
+    private static final String DEVICE_DATA_FILE_PATH = FileManager.getDeviceConfigPath() + Constant.DEVICE_DATA_FILE;
+
+    @PostConstruct
+    private void init() {
+        ObjectInputStream objectInputStream = null;
+        try {
+            File file = new File(DEVICE_DATA_FILE_PATH);
+            if (!file.exists()) {
+                return;
+            }
+            objectInputStream = new ObjectInputStream(new FileInputStream(file));
+            Map<String, DeviceInfo> deviceInfoMap = (Map<String, DeviceInfo>) objectInputStream.readObject();
+            reLogin(deviceInfoMap);
+            CACHE_DEVICE_INFO.putAll(deviceInfoMap);
+        } catch (Exception e) {
+            logger.warn(e.getMessage(),e);
+        } finally {
+            quietClose(objectInputStream);
+        }
+    }
+
+    @Override
+    public Map<String, DeviceInfo> getAll() {
+        return CACHE_DEVICE_INFO;
+    }
+
+    @Override
+    public void remove(String uid) {
+        DeviceInfo deviceInfo = CACHE_DEVICE_INFO.remove(uid);
+        LOGINED_DEVICE.remove(deviceInfo.ip);
+        writeDataToFile();
+    }
+
+    @Override
+    public void update(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+    @Override
+    public void add(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+
+    @Override
+    public DeviceInfo get(String uid) {
+        return CACHE_DEVICE_INFO.get(uid);
+    }
+
+    @Override
+    public Storage getLoginedDeviceByIP(String deviceIP) {
```

Comment:
better to rename the function as getLoggedInDeviceByIP()

Reply:
Yes, Thanks

Reply:
Done